PS-3767: Fixed clang 4/5 warnings.

dutow opened the pull request:
* Removed unneccessary parantheses from conditions
* Disabled the tautological constant out of range warning (static asserts would require c++11)
* Fixed a const char* -> char* conversion
* Changed null checks of arrays to compile time assertions that they are arrays
* Added some explicit casts
* Removed unused functions / variables
* Marked functions which are only used in some configurations as possibly unused

laurynas-biveinis commented on cmd-line-utils/readline/history.c:
```diff
@@ -211,14 +211,14 @@ history_get (offset)
   temp = (HIST_ENTRY *)xmalloc (sizeof (HIST_ENTRY));
 
-  temp->line = string ? savestring (string) : string;
+  temp->line = string ? savestring (string) : NULL;
```
What was the warning here?

dutow replied:
This is part of the other change in this file: string was a char*, and not a const char*. And line is simply a char* (and has to stay that), so we can't assign string directly to it, even when we know it's null.

laurynas-biveinis commented on libmysql/libmysql.c:
```diff
@@ -1379,7 +1379,8 @@ void set_stmt_errmsg(MYSQL_STMT *stmt, NET *net)
 
   stmt->last_errno= net->last_errno;
-  if (net->last_error && net->last_error[0])
+  compile_time_assert(sizeof(net->last_error)/sizeof(void*) > 1);
```
Spaces around /

laurynas-biveinis commented on sql/item.cc:
```diff
@@ -7035,7 +7035,7 @@ bool Item_default_value::fix_fields(THD *thd, Item **items)
   if (!(def_field= (Field*) sql_alloc(field_arg->field->size_of())))
     goto error;
-  memcpy(def_field, field_arg->field, field_arg->field->size_of());
+  memcpy((void*)def_field, (void*)field_arg->field, field_arg->field->size_of());
```
What was the warning fixed here?

dutow replied:
memcpy is used to copy a class with virtual functions, and clang generates the warning that in this case, it'll copy the vtblptr. Which is fine, but this is the way the warning can be silenced.

laurynas-biveinis replied:
Isn't memcpy on such object undefined behavior, i.e. not fine? What do higher versions do?

dutow replied:
It is, but I think in practice it works in every implementation. 

They refactored this part, by introducing a pure virtual clone method in the Field class, and implementing it in every derived class.

laurynas-biveinis replied:
OK, that's too much to backport, let's just add the typecasts. Please use C++ style for them

laurynas-biveinis commented on sql/log.cc:
```diff
@@ -1236,8 +1236,9 @@ bool LOGGER::slow_log_print(THD *thd, const char *query, uint query_length,
     }
 
     /* fill in user_host value: the format is "%s[%s] @ %s [%s]" */
+    compile_time_assert(sizeof(sctx->priv_user)/sizeof(void*)>1);
```
Spaces around operators

laurynas-biveinis commented on sql/log_event.cc:
```diff
@@ -1164,7 +1164,7 @@ failed my_b_read"));
 #ifndef max_allowed_packet
   THD *thd=current_thd;
-  uint max_allowed_packet= thd ? slave_max_allowed_packet:~(ulong)0;
+  uint max_allowed_packet= thd ? slave_max_allowed_packet:~(uint)0;
```
Spaces around :

laurynas-biveinis commented on sql/log_event.cc:
```diff
@@ -2400,8 +2400,9 @@ bool Query_log_event::write(IO_CACHE* file)
-    else if (thd->security_ctx->priv_user)
+    else
     {
+      compile_time_assert(sizeof(thd->security_ctx->priv_user)/sizeof(void*)>1);
```
Spaces around operators

laurynas-biveinis commented on sql/slave.cc:
```diff
@@ -4099,8 +4099,8 @@ static int queue_event(Master_info* mi,const char* buf, ulong event_len)
     */
-    if ((memcmp(mi->master_log_name, hb.get_log_ident(), hb.get_ident_len())
-         && mi->master_log_name != NULL)
+    compile_time_assert(sizeof(mi->master_log_name)/sizeof(void*)>1);
```
Likewise

laurynas-biveinis commented on sql/slave.cc:
```diff
@@ -4358,7 +4358,8 @@ static int connect_to_master(THD* thd, MYSQL* mysql, Master_info* mi,
 
   /* we disallow empty users */
-  if (mi->user == NULL || mi->user[0] == 0)
+  compile_time_assert(sizeof(mi->user)/sizeof(void*)>1);
```
Likewise

laurynas-biveinis commented on sql/slave.cc:
```diff
@@ -4494,8 +4495,8 @@ MYSQL *rpl_connect_master(MYSQL *mysql)
 
-  if (mi->user == NULL
-      || mi->user[0] == 0
+  compile_time_assert(sizeof(mi->user)/sizeof(void*)>1);
```
Likewise

laurynas-biveinis commented on sql/sql_acl.cc:
```diff
@@ -558,22 +558,22 @@ TABLE_FIELD_TYPE mysql_tables_priv_table_fields[MYSQL_TABLES_PRIV_FIELD_COUNT] =
   mysql_db_table_def= {MYSQL_DB_FIELD_COUNT, mysql_db_table_fields};
 
-const TABLE_FIELD_DEF
+__attribute__((unused)) const TABLE_FIELD_DEF
```
In what configurations this gets skipped? NO_EMBEDDED_ACCESS_CHECKS? Can't you just use the same ifdef here instead of marking as unused?

If fixed in the current way, then MY_ATTRIBUTE everywhere

dutow replied:
MY_ATTRIBUTE was introduced in 5.6, 5.5 used simply attribute (there is an ifdef for it in include/my_attribute.h)

Yes, it's that definition, I could add that ifdef here too.

laurynas-biveinis replied:
Please use that definition (but my comment of taking higher version fixes, if available, takes precedence)

laurynas-biveinis commented on sql/sql_acl.cc:
```diff
@@ -5719,7 +5719,8 @@ bool check_grant_db(THD *thd,const char *db)
   size_t copy_length;
 
-  copy_length= (size_t) (strlen(sctx->priv_user ? sctx->priv_user : "") +
+  compile_time_assert(sizeof(sctx->priv_user)/sizeof(void*)>1);
```
Spaces around operators

laurynas-biveinis replied:
Comment still stands

laurynas-biveinis commented on sql/sql_acl.cc:
```diff
@@ -10397,10 +10392,10 @@ acl_authenticate(THD *thd, uint connect_errors, uint com_change_user_pkt_len)
   {
     if (strcmp(mpvio.auth_info.authenticated_as, mpvio.auth_info.user_name))
     {
+      compile_time_assert(sizeof(mpvio.auth_info.authenticated_as)/sizeof(void*)>1);
```
Likewise

laurynas-biveinis commented on sql/sql_audit.cc:
```diff
@@ -120,7 +120,7 @@ static audit_handler_t audit_handlers[] =
 };
 
-static const uint audit_handlers_count=
+__attribute__((unused)) static const uint audit_handlers_count=
```
Same comments as for the previous unused attribute

laurynas-biveinis commented on sql/sql_audit.cc:
```diff
@@ -345,8 +345,8 @@ int initialize_audit_plugin(st_plugin_int *plugin)
   
-  if (!data->class_mask || !data->event_notify ||
-      !data->class_mask[0])
+  compile_time_assert(sizeof(data->class_mask)/sizeof(void*)>0);
```
Spaces

laurynas-biveinis commented on sql/sql_parse.cc:
```diff
@@ -4031,8 +4031,8 @@ case SQLCOM_PREPARE:
       goto error;
-    if ((thd->security_ctx->priv_user &&
-	 !strcmp(thd->security_ctx->priv_user, grant_user->user.str)) ||
+    compile_time_assert(sizeof(thd->security_ctx->priv_user)/sizeof(void*)>1);
```
Spaces

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -1938,6 +1938,7 @@ trx_is_registered_for_2pc(
 
 /*********************************************************************//**
 Note that a transaction owns the prepare_commit_mutex. */
+__attribute__((__unused__))
```
Is this also used in some configs only?

dutow replied:
Yes, only if EXTENDED_FOR_COMMIT_ORDERED

laurynas-biveinis replied:
OK, let's use it then. (I suspect that this define is ours, not upstream, and we could remove it, but 5.5 is a mature release next in line for EOL, so maybe let's just fix the warning)

laurynas-biveinis commented:
If taking an upstream fix from a higher version, and that fix commit scope matches, cherry picks are OK (then this PR can have e.g. several cherry picks and own commit on the top for the rest)

dutow commented:
Updated the PR.

I was only able to use one of the upstream commits - the other is focused on warning fixes too, but there were more conflicts than useful changes in it.

laurynas-biveinis commented on cmake/maintainer.cmake:
```diff
@@ -31,7 +31,7 @@ SET(MY_CXX_WARNING_FLAGS
 IF(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
   SET(MY_CXX_WARNING_FLAGS
-      "${MY_CXX_WARNING_FLAGS} -Wno-null-conversion -Wno-unused-private-field")
+      "${MY_CXX_WARNING_FLAGS} -Wno-null-conversion -Wno-unused-private-field -Wno-tautological-constant-out-of-range-compare")
```
On a second though, please add a comment here so this is removed in an eventual merge to 8.0. This is not required if 5.6 or 5.7 already don't need this

dutow replied:
This is only required for some assertions: the assertion checks that the actual value (which has an enum type) returned by a function is less than 256. Clang sees that the largest value of the enum is smaller than 256 (as it should be), and reports a warning because of this.

I didn't see this warning in later versions, so  I think these asserts were removed or changed.

laurynas-biveinis replied:
I misunderstood the PR description: I thought it is the compile_time_assert itself which caused warnings here.

In this case it is those higher version fixes we should take, unless they are intertwined with other major works. It is also an option to remove the tautological asserts.

dutow replied:
That's a similar, but different warnings, and that was something I could fix with compile-time assertions. 

I wanted to avoid the removal of these asserts, as they are only tautological until they do work - the moment they break, we wouldn't receive this warning. (when somebody adds a value greater than 256 to these enums)

dutow replied:
I updated the PR with a version where I removed the two related asserts (unireg.cc and ha_innodb.cc)

laurynas-biveinis replied:
Would a cast, inside the assert, to a wider type before comparing work?

laurynas-biveinis commented on sql/log.cc:
```diff
@@ -1236,8 +1236,9 @@ bool LOGGER::slow_log_print(THD *thd, const char *query, uint query_length,
     }
 
     /* fill in user_host value: the format is "%s[%s] @ %s [%s]" */
+    compile_time_assert(sizeof(sctx->priv_user) / sizeof(void*)>1);
```
Spaces around >

laurynas-biveinis commented on sql/log_event.cc:
```diff
@@ -2400,8 +2400,9 @@ bool Query_log_event::write(IO_CACHE* file)
-    else if (thd->security_ctx->priv_user)
+    else
     {
+      compile_time_assert(sizeof(thd->security_ctx->priv_user) / sizeof(void*)>1);
```
Likewise

laurynas-biveinis commented on sql/slave.cc:
```diff
@@ -4099,8 +4099,8 @@ static int queue_event(Master_info* mi,const char* buf, ulong event_len)
     */
-    if ((memcmp(mi->master_log_name, hb.get_log_ident(), hb.get_ident_len())
-         && mi->master_log_name != NULL)
+    compile_time_assert(sizeof(mi->master_log_name) / sizeof(void*)>1);
```
Likewise

laurynas-biveinis commented on sql/slave.cc:
```diff
@@ -4358,7 +4358,8 @@ static int connect_to_master(THD* thd, MYSQL* mysql, Master_info* mi,
 
   /* we disallow empty users */
-  if (mi->user == NULL || mi->user[0] == 0)
+  compile_time_assert(sizeof(mi->user) / sizeof(void*)>1);
```
Likewise

laurynas-biveinis commented on sql/slave.cc:
```diff
@@ -4494,8 +4495,8 @@ MYSQL *rpl_connect_master(MYSQL *mysql)
 
-  if (mi->user == NULL
-      || mi->user[0] == 0
+  compile_time_assert(sizeof(mi->user) / sizeof(void*)>1);
```
Likewise

laurynas-biveinis commented on sql/sql_acl.cc:
```diff
@@ -576,6 +577,7 @@ const TABLE_FIELD_DEF
 const TABLE_FIELD_DEF
   mysql_tables_priv_table_def= {MYSQL_TABLES_PRIV_FIELD_COUNT,
                                 mysql_tables_priv_table_fields};
+#endif //NO_EMBEDDED_ACCESS_CHECKS
```
Space after //

laurynas-biveinis commented on sql/sql_acl.cc:
```diff
@@ -10397,10 +10394,10 @@ acl_authenticate(THD *thd, uint connect_errors, uint com_change_user_pkt_len)
   {
     if (strcmp(mpvio.auth_info.authenticated_as, mpvio.auth_info.user_name))
     {
+      compile_time_assert(sizeof(mpvio.auth_info.authenticated_as) / sizeof(void*)>1);
```
Spaces around >

laurynas-biveinis commented on sql/sql_audit.cc:
```diff
@@ -120,9 +120,10 @@ static audit_handler_t audit_handlers[] =
   general_class_handler, connection_class_handler
 };
 
+#if !defined(DBUG_OFF) && !defined(_lint)
```
Why _lint too?

dutow replied:
I copied this ifdef from the definition of the assert macro, which uses the variable.

laurynas-biveinis commented on sql/sql_audit.cc:
```diff
@@ -345,8 +346,8 @@ int initialize_audit_plugin(st_plugin_int *plugin)
   
-  if (!data->class_mask || !data->event_notify ||
-      !data->class_mask[0])
+  compile_time_assert(sizeof(data->class_mask) / sizeof(void*)>0);
```
Spaces around >

laurynas-biveinis commented on sql/sql_parse.cc:
```diff
@@ -4031,8 +4031,8 @@ case SQLCOM_PREPARE:
       goto error;
-    if ((thd->security_ctx->priv_user &&
-	 !strcmp(thd->security_ctx->priv_user, grant_user->user.str)) ||
+    compile_time_assert(sizeof(thd->security_ctx->priv_user) / sizeof(void*)>1);
```
Likewise

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -1947,6 +1948,7 @@ trx_owns_prepare_commit_mutex_set(
 	ut_a(trx_is_registered_for_2pc(trx));
 	trx->owns_prepare_mutex = 1;
 }
+#endif //EXTENDED_FOR_COMMIT_ORDERED
```
Space after //

dutow commented:
New jenkins run: https://jenkins.percona.com/job/percona-server-5.5-param/1607/

laurynas-biveinis commented on sql/unireg.cc:
```diff
@@ -924,7 +924,6 @@ static bool pack_fields(File file, List<Create_field> &create_fields,
     recpos= field->offset+1 + (uint) data_offset;
     int3store(buff+5,recpos);
     int2store(buff+8,field->pack_flag);
-    DBUG_ASSERT(field->unireg_check < 256);
```
IMHO what this is trying to say is compile_time_assert(sizeof() == 1)

dutow replied:
Yes, that would be the intention, but that check would fail - because the size of that field is 4 byte, not 1 - it just never contains a value higher than 256.

percona-ysorokin commented:
@dutow Please also double check if there are no build issues with clang when both `-DCMAKE_BUILD_TYPE=RelWithDebInfo` and `-DMYSQL_MAINTAINER_MODE=ON` are enabled.

dutow commented:
@percona-ysorokin The relwithdebinfo errors aren't clang specific, I'll submit those in a separate PR.

dutow commented:
A small update:
* rebased to latest 5.5 head
* added the llvm-dev package to the travis config
* fixed a header guard in the bundled ssl library

With these, the travis check is green.

laurynas-biveinis commented on :q:
```diff
@@ -0,0 +1,77 @@
+# Ubuntu 14.04
```
File named ":q"? Switch to Emacs ASAP!

dutow replied:
Ops, removed :)

laurynas-biveinis commented on extra/yassl/include/openssl/crypto.h:
```diff
@@ -19,7 +19,7 @@
 /* crypto.h for openSSL */
 
-#ifndef ysSSL_crypto_h__
+#ifndef yaSSL_crypto_h__
```
Please report upstream if present in 5.6 at least

dutow replied:
It isn't. The yassl versions is different in the 3 versions, and I've only seen (different) issues in 5.5 and 5.7

laurynas-biveinis commented:
Now that we have switched, s/LP/JIRA in the commit message

dutow commented:
Commit message / pr name edited, file removed.
